rxjava2-jdbc/src/main/java/org/davidmoten/rx/jdbc/Util.java

```diff
@@ -287,39 +287,41 @@ static void closeCallableStatementAndConnection(NamedCallableStatement stmt) {
         closePreparedStatementAndConnection(stmt.stmt);
     }
 
-    static NamedPreparedStatement prepare(Connection con, String sql) throws SQLException {
-        return prepare(con, 0, sql);
+    static NamedPreparedStatement prepare(Connection con, String sql, int queryTimeoutSec) throws SQLException {
+        return prepare(con, 0, sql, queryTimeoutSec);
     }
 
-    static NamedPreparedStatement prepare(Connection con, int fetchSize, String sql) throws SQLException {
+    static NamedPreparedStatement prepare(Connection con, int fetchSize, String sql, int queryTimeoutSec) throws SQLException {
         // TODO can we parse SqlInfo through because already calculated by
         // builder?
         SqlInfo info = SqlInfo.parse(sql);
         log.debug("preparing statement: {}", sql);
-        return prepare(con, fetchSize, info);
+        return prepare(con, fetchSize, info, queryTimeoutSec);
     }
 
-    static PreparedStatement prepare(Connection connection, int fetchSize, String sql, List<Parameter> parameters)
+    static PreparedStatement prepare(Connection connection, int fetchSize, String sql, List<Parameter> parameters, int queryTimeoutSec)
             throws SQLException {
         // should only get here when parameters contains a collection
         SqlInfo info = SqlInfo.parse(sql, parameters);
         log.debug("preparing statement: {}", info.sql());
-        return createPreparedStatement(connection, fetchSize, info);
+        return createPreparedStatement(connection, fetchSize, info, queryTimeoutSec);
     }
 
-    private static NamedPreparedStatement prepare(Connection con, int fetchSize, SqlInfo info) throws SQLException {
-        PreparedStatement ps = createPreparedStatement(con, fetchSize, info);
+    private static NamedPreparedStatement prepare(Connection con, int fetchSize, SqlInfo info, int queryTimeoutSec) throws SQLException {
+        PreparedStatement ps = createPreparedStatement(con, fetchSize, info, queryTimeoutSec);
         return new NamedPreparedStatement(ps, info.names());
     }
 
-    private static PreparedStatement createPreparedStatement(Connection con, int fetchSize, SqlInfo info)
+    private static PreparedStatement createPreparedStatement(Connection con, int fetchSize, SqlInfo info, int queryTimeoutSec)
             throws SQLException {
         PreparedStatement ps = null;
         try {
             ps = con.prepareStatement(info.sql(), ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY);
             if (fetchSize > 0) {
                 ps.setFetchSize(fetchSize);
             }
+
+            ps.setQueryTimeout(queryTimeoutSec);
```

Comment:
We haven't covered the case where a default query timeout has been set in the connection string say of 20s yet we will override it to 0 unless explicitly specified in each statement. I think what we need is a default value for queryTimeoutSec of -1 which means don't set it.

Reply:
Note also that I wouldn't suggest allowing the user to set a query timeout of -1, it should just be the default if they don't set it.

Reply:
I think that according to the JDBC Statement interface, 0 is the correct default when you don't want any timeout:

  * @param seconds the new query timeout limit in seconds; zero means
     *        there is no limit
     * @exception SQLException if a database access error occurs,
     * this method is called on a closed <code>Statement</code>
     *            or the condition {@code seconds >= 0} is not satisfied
     * @see #getQueryTimeout
     */
    void setQueryTimeout(int seconds) throws SQLException;

Reply:
Ah no, I meant -1 for not setting the timeout at all so that a default timeout is applied.